### ISSUE TITLE
Feature Request: multi-threaded uploading/downloading

### ISSUE THREAD

**Issue by priv** [abraunegg/onedrive#232]
Original OneDrive Windows client creates up to 4 threads so that uploading small files would be faster.
Is it possible to implement this feature?  Is there API limitation and/or program architecture limitations?

**Comment by abraunegg** [comment]
@priv 
I have been looking into this for quite a while - and the current roadblocks are this:

1. console logging output of operations / logfile output needs to be totally re-written so that individual functions can fire off the 'output' and keep processing, but the logging engine handles a buffer & FIFO for writing the entries out in the order they were received. Issue here obviously is the 'download progress bar' for files larger than X - how to handle in a multi threaded environment.

2. The current functions as written / classes don't allow themselves to be easily run in a multi-threaded fashion (i've already been looking on how to do this). 

In short - nothing short of an entire re-write will enable this to work properly - which will take quite a bit of planning and engineering / time.

**Comment by abraunegg** [comment]
Linked with #157

**Comment by kaszperro** [comment]
Is there any progress in terms of this issue?

**Comment by abraunegg** [comment]
@kaszperro 
High level mapping has been done, however my actual work takes precedence over this.

**Comment by cgeopapa** [comment]
Any progress on this feature?

**Comment by Gby56** [comment]
Any hope for this improvement to be implemented ? I don't know much D programming but I saw https://github.com/abraunegg/onedrive/pull/630 , or would https://github.com/onedrivejs/onedrive be recommended ?

**Comment by abraunegg** [comment]
@Gby56 
It is still a work in progress, which is being worked on, but bugs / issues take priority.

That Javascript client has issues as well with threading / handling / support of account types / support of nuances of how OneDrive API supports file types ....

**Comment by JordanStopford** [comment]
Happy to test this when it's available. We have a sharepoint with around 2 million files and the synchronization is rather slow (still downloading and has been going for 6 days)

**Comment by sclsj** [comment]
Any progress? I have a Synology NAS; while Cloud Sync does support concurrent download/upload, it does not support OneDrive Business with external MFA (which is what I have), so I have to look for other OneDrive clients.

**Comment by skupjoe** [comment]
Looking forward to this! 👍 
I sync a lot of git repos (bad practice, I know..but it's important to me) and I needed to switch from Dropbox linux client to this.
Dropbox client seems to handle multithreading pretty well, so eager to have this supported with OneDrive as well

**Comment by JL102** [comment]
Keywords for search: Multiple downloads, concurrent, concurrency, parallel, speed (I had difficulty finding this thread earlier because I didn't have the right search terms)

**Comment by JL102** [comment]
Question: If and when this is implemented, would it also speed up the "Processing x items to ensure consistent local state" stage? It takes forever and a half for the OneDrive client to start up as it goes through that process.

**Comment by abraunegg** [comment]
> Question: If and when this is implemented, would it also speed up the "Processing x items to ensure consistent local state" stage? It takes forever and a half for the OneDrive client to start up as it goes through that process.

Please raise a question in the right place: https://github.com/abraunegg/onedrive/discussions

**Comment by abraunegg** [comment]
This feature is in-progress development. Refer to https://github.com/abraunegg/onedrive/discussions/2415 for tracking details

**Comment by abraunegg** [comment]
This feature has been developed and needs extensive testing. Refer to https://github.com/abraunegg/onedrive/discussions/2478 for details.

**Comment by abraunegg** [comment]
Marking as Fixed | Implemented as this is included in v2.5.0 alpha series codebase.
